finality-aleph/src/data_io.rs

```diff
@@ -13,73 +13,256 @@ use futures::{
+    collections::{BTreeMap, HashMap, HashSet},
+    hash::{Hash, Hasher},
     marker::PhantomData,
     sync::Arc,
     time::{self, Duration},
 };
 
 type MessageId = u64;
+pub(crate) const MAX_DATA_BRANCH_LEN: u32 = 7;
 const REFRESH_INTERVAL: u64 = 100;
-const AVAILABLE_BLOCKS_CACHE_SIZE: usize = 1000;
-const MESSAGE_ID_BOUNDARY: MessageId = 100_000;
+const AVAILABLE_BLOCKS_CACHE_SIZE: usize = 2000;
+const MAX_MESSAGES_PENDING: usize = 40_000;
+const MAX_PROPOSALS_PENDING: usize = 80_000;
 const PERIODIC_MAINTENANCE_INTERVAL: Duration = Duration::from_secs(60);
 const REQUEST_FORK_AFTER: Duration = Duration::from_secs(100);
 
 /// The data ordered by the Aleph consensus.
-#[derive(Copy, PartialEq, Eq, Clone, Debug, Encode, Decode, Hash)]
-pub struct AlephData<H, N> {
-    pub hash: H,
-    pub number: N,
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum AlephData<B: BlockT> {
+    Empty,
+    HeadProposal(AlephProposal<B>),
 }
 
-impl<H, N> AlephData<H, N> {
-    pub(crate) fn new(block_hash: H, block_number: N) -> Self {
-        AlephData {
-            hash: block_hash,
-            number: block_number,
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephData<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            AlephData::Empty => {
+                (0u8).hash(state);
+            }
+            AlephData::HeadProposal(proposal) => {
+                (1u8).hash(state);
+                proposal.hash(state);
+            }
+        }
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephData<B> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (AlephData::Empty, AlephData::Empty) => true,
+            (AlephData::HeadProposal(p1), AlephData::HeadProposal(p2)) => p1.eq(p2),
+            _ => false,
         }
     }
 }
 
-pub(crate) type AlephDataFor<B> = AlephData<<B as BlockT>::Hash, NumberFor<B>>;
+impl<B: BlockT> Eq for AlephData<B> {}
+
+impl<B: BlockT> AlephData<B> {
+    pub(crate) fn is_correct(&self) -> bool {
+        match self {
+            AlephData::Empty => true,
+            AlephData::HeadProposal(proposal) => proposal.is_correct(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct AlephProposal<B: BlockT> {
+    // We expect that honest nodes create AlephData {branch: [h_0, h_1, ..., h_n], number: num} objects that represent
+    // an ascending sequence of blocks b_0, b_1, ..., b_n satisfying the following conditions:
+    //     1) hash(b_i) = h_i for i = 0, 1, ..., n,
+    //     2) parent(b_{i+1}) = b_i for i = 0, 1, ..., (n-1),
+    //     3) height(b_n) = num,
+    //     4) The parent of b_0 has been finalized (prior to creating this AlephData).
+    // Such an AlephData object should be thought of as a proposal for block b_n to be finalized.
+    pub branch: Vec<B::Hash>,
```

Comment:
I would consider explicitly bounded data structure like `BoundedVec`

Reply:
Generally that makes some sense, but:
1) My understanding is that `BoundedVec` is only a wrapping for `Vec` that is somehow optimized for substrate storage. So I don't think there would be any benefit in using it here.
2) I made a quick search online and didn't find anything worth attention. Also we would likely need external dependencies for that.